lib/model/model.go

```diff
@@ -1372,9 +1372,64 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
 	} else if err != nil {
 		return protocol.ErrGeneric
 	}
+
+	if !scanner.Validate(buf, hash, weakHash) {
+		m.recheckFile(deviceID, folderFs, folder, name, int(offset/protocol.BlockSize), hash)
+	}
+
 	return nil
 }
 
+func (m *Model) recheckFile(deviceID protocol.DeviceID, folderFs fs.Filesystem, folder, name string, blockIndex int, hash []byte) {
+	cf, ok := m.CurrentFolderFile(folder, name)
+	if !ok {
+		l.Debugf("%v recheckFile: %s: %q / %q: no current file", m, deviceID, folder, name)
+		return
+	}
+
+	if cf.IsDeleted() || cf.IsInvalid() || cf.IsSymlink() || cf.IsDirectory() {
+		l.Debugf("%v recheckFile: %s: %q / %q: not a regular file", m, deviceID, folder, name)
+		return
+	}
+
+	if blockIndex > len(cf.Blocks) {
+		l.Debugf("%v recheckFile: %s: %q / %q i=%d: block index too far", m, deviceID, folder, name, blockIndex)
+		return
+	}
+
+	block := cf.Blocks[blockIndex]
+
+	// Seems to want a different version of the file, whatever.
+	if !bytes.Equal(block.Hash, hash) {
+		l.Debugf("%v recheckFile: %s: %q / %q i=%d: hash mismatch %x != %x", m, deviceID, folder, name, blockIndex, block.Hash, hash)
+		return
+	}
+
+	if blocks, err := scanner.HashFile(context.Background(), folderFs, name, protocol.BlockSize, nil, false); err == nil {
+		if protocol.BlocksEqual(cf.Blocks, blocks) {
+			l.Debugf("%v recheckFile: %s: %q / %q: blocks match", m, deviceID, folder, name)
+			return
+		} else {
+			l.Debugf("%v recheckFile: %s: %q / %q: blocks do not match", m, deviceID, folder, name, err)
+		}
+	} else {
+		l.Debugf("%v recheckFile: %s: %q / %q: error while hashing file: %s", m, deviceID, folder, name, err)
+	}
+
+	// The file probably did not get rescanned, as mtimes haven't changed.
+	// Invalidate it, and trigger another scan.
+	cf.Invalidate(m.shortID)
```

Comment:
I don't get where mtime comes into play. Neither do I understand the need to invalidate (that's maybe linked to not getting the mtime bit).  
In general wouldn't it be simpler and use less resources to rescan the file directly instead of explicitly hashing after checking the block hash. The current flow hashes a changed file twice.

Reply:
The idea here is as follows:
1. We get what other side expects to get as part of the request.
2. We validate that what we've read matches the weak hash, if it does, nothing happens.
3. If it doesn't match, we validate using SHA256, if it matches it's all good.
4. If it still doesn't match, look up our own file in the database, and compare hashes for that block. If the our hashes don't match request hashes, it means the requestor is asking for something different, so we do nothing.


You are right that we don't need to hash here, as we know that what we've read doesn't match the hash the user has asked us for, and we also know that our hash matches his, which means the file is somehow changed, so we can skip the HashFile and go straight to invalidate and rescan.

The idea behind the mtime is the fact that if something modifies files in place, or artificially adjusts mtimes, we could miss some changes, hence we double check what we're reading.

Reply:
Ok, I get the invalidation now. One little nick there: In almost all cases it will result in an index update sent with the invalid file info directly followed by an update valid file info. Only in the rare case of a file that has unchanged size and mtime, but changed content, will the invalidation have any effect, in all others it is useless overhead. You could pass the file info from `Request` and check whether it is equivalent to the current file in the database (as in `(f FileInfo) IsEquivalent`)  and invalidate if it is and scan otherwise.

Reply:
Request does not have a FileInfo, it just has folder, name, offset, size, hash, weakHash.
Essentially the goal is to force rescan a file.

Reply:
I don't know whether force rescanning is a good idea. The only real case where a  file is has changed content without changed size&mtime I can think of is bitrot. In that case we'd rather not scan and send the data.

As to what I meant with checking whether the file is changed: Within `Request` the file is stated and within `recheckFile` the current file info is obtained from the db. That's all the information needed to check whether the two are "equivalent".